### PR TITLE
Fix output types of transformed values onSubmit

### PR DESCRIPTION
**Pull request by christopherklint97**
## Summary

This fixes a typing/validation issue when using a resolver that transforms the form input types into different output types.

Example using zod:
```typescript

import { useForm } from "react-hook-form";
import { z } from "zod";
import { zodResolver } from "@hookform/resolvers/zod";

const countSchema = z.object({
  count: z.string().transform(val => Number(val)),
});

type CountForm = z.input<typeof countSchema>; // input type is inferred => string
type CountSchema = z.infer<typeof countSchema>; // output type is inferred => number

export default function App() {
  const { register, handleSubmit } = useForm<CountForm>({resolver: zodResolver(countSchema)});
  const onSubmit = (data: CountSchema) => console.log(data);

  return (
    <div>
      ...
      // Error: Type '{ count: string; }' is not assignable to type '{ count: number; }'.
      <form onSubmit={handleSubmit(onSubmit)}>
        <input {...register("count")} placeholder="Count" />
        <input type="submit" />
      </form>
    </div>
  );
}
```
Error was also reproduced here: https://codesandbox.io/s/ecstatic-pine-brxt1?file=/src/App.tsx:0-892

### Version

7.26.1

### CODE REVIEW

File: src/__tests__/useForm/handleSubmit.test.tsx [hunk 1]
```diff
@@ -349,6 +349,34 @@ describe('handleSubmit', () => {
       });
       expect(callback.mock.calls[0][0]).toEqual({ test: 'test' });
     });
+
+    it('should transform input values and return correct output types', async () => {
```
Comment (bluebill1049) [inline src/__tests__/useForm/handleSubmit.test.tsx:353]: Thanks for the PR. I don't think this test does cover the type fix as it passes today's codebase.

File: reports/api-extractor.md [hunk 2]
```diff
@@ -550,7 +550,7 @@ export type UseFormGetValues<TFieldValues extends FieldValues> = {
 };
 
 // @public
-export type UseFormHandleSubmit<TFieldValues extends FieldValues> = <_>(onValid: SubmitHandler<TFieldValues>, onInvalid?: SubmitErrorHandler<TFieldValues>) => (e?: React_2.BaseSyntheticEvent) => Promise<void>;
```
Comment (bluebill1049) [inline reports/api-extractor.md:553]: `<_>` was left there to avoid breaking change.

File: src/__tests__/useForm/handleSubmit.test.tsx (hunk 1 above)
Comment (christopherklint97) [reply src/__tests__/useForm/handleSubmit.test.tsx:353]: Thank you for your comments! You were right about that, I change the test so it should be failing now.

File: reports/api-extractor.md (hunk 2 above)
Comment (christopherklint97) [reply reports/api-extractor.md:553]: Ahh ok, then I wonder if this is the intentional behavior of react-hook-forms, that the output values of the resolver should be the `TFieldValues` and not the input values if they are transformed? 
For example:
```typescript
import { useForm } from "react-hook-form";
import { z } from "zod";
import { zodResolver } from "@hookform/resolvers/zod";

const countSchema = z.object({
  count: z.string().transform(val => Number(val)),
});

type CountSchema = z.infer<typeof countSchema>; // output type is inferred => number

export default function App() {
  const { register, handleSubmit } = useForm<CountSchema>({resolver: zodResolver(countSchema)});
  const onSubmit = (data: CountSchema) => console.log(data);
  ...
```

Been trying to think of other ways to transform the typing but come up short 🤷🏾

**Comment by Moshyfawn** [comment]
Does #7730 help?

**Comment by christopherklint97** [comment]
> Does #7730 help?

Yes it does help definitely, thanks! I think I misunderstood the problem and thought that the resolvers should notify react-hook-form somehow about the types of the return values.

I guess it is more of an issue with zod's paradigm then, that values can be transformed to new types after being parsed.

Really unsure though how to type the `useForm` and `data` object from `handleSubmit` since the input values are the only things being considered from either the `defaultValues` or the generic type in `useForm<T>` 🤔

**Pull request closed without merging**